### PR TITLE
Add MC reconstructor to advection subcell

### PR DESCRIPTION
**Pull request by yoonso0-0**
## Proposed changes
* Add `ReconstructWork` 
* Add monotised-central reconstructor for advection system
- [x] ~~#3532~~

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

**Comment by nilsdeppe** [comment]
@yoonso0-0 this needs a rebase now that the other PR is merged

File: src/Evolution/Systems/ScalarAdvection/FiniteDifference/ReconstructWork.hpp [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+class not_null;
+}  // namespace gsl
+/// \endcond
+
+namespace ScalarAdvection::fd {
+/*!
+ * \brief Reconstructs the scalar field \f$U\f$. All results are written into
+ * `vars_on_lower_face` and `vars_on_upper_face`.
+ *
+ * Need more detailed docs - direction, lower and upper face?
```
Comment (nilsdeppe) [inline src/Evolution/Systems/ScalarAdvection/FiniteDifference/ReconstructWork.hpp]: I would say hold off on writing more until we see what unstructured mesh reconstruction looks like :)

File: src/Evolution/Systems/ScalarAdvection/FiniteDifference/ReconstructWork.hpp [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+                       evolution::dg::subcell::NeighborData,
+                       boost::hash<std::pair<Direction<Dim>, ElementId<Dim>>>>
+        neighbor_data,
+    const Mesh<Dim>& subcell_mesh, const size_t ghost_zone_size) noexcept;
+
+/*!
+ * \brief Reconstructs the scalar field \f$U\f$ for a given direction. All
+ * results are written into `vars_on_face`.
+ *
+ * This is used on DG elements to reconstruct boundary correction on the element
```
Comment (nilsdeppe) [inline src/Evolution/Systems/ScalarAdvection/FiniteDifference/ReconstructWork.hpp]: It doesn't reconstruction the boundary correction, it reconstructs `U` and then computes the neighbor's input into the boundary correction. The order of when you do reconstruction matters, so we need to be explicit. Why not use the dox from GRMHD? You could expand it, but that dox is technically correct

File: src/Evolution/Systems/ScalarAdvection/FiniteDifference/ReconstructWork.hpp [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+ * interface based on their subcell neighbors' solution (ghost cells received
+ * from the FD neighbor)
+ */
+template <size_t Dim, typename TagsList, typename Reconstruct_lower,
```
Comment (nilsdeppe) [inline src/Evolution/Systems/ScalarAdvection/FiniteDifference/ReconstructWork.hpp]: `Reconstruct_lower` -> `ReconstructLower` (same with upper). These are type names

File: src/Evolution/Systems/ScalarAdvection/FiniteDifference/ReconstructWork.tpp [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+                       evolution::dg::subcell::NeighborData,
+                       boost::hash<std::pair<Direction<Dim>, ElementId<Dim>>>>
+        neighbor_data,
+    const Mesh<Dim>& subcell_mesh, const size_t ghost_zone_size) {
+  // check if subcell mesh is isotropic
+  ASSERT(Mesh<Dim>(subcell_mesh.extents(0), subcell_mesh.basis(0),
+                   subcell_mesh.quadrature(0)) == subcell_mesh,
+         "The subcell mesh should be isotropic but got " << subcell_mesh);
+
+  // number of points at which reconstruction is performed for each dimension,
```
Comment (nilsdeppe) [inline src/Evolution/Systems/ScalarAdvection/FiniteDifference/ReconstructWork.tpp]: I'm not sure what `at which reconstruction` means. I would say `The number of interface points, which is equal to the number of points we reconstruct to`. Similar to interpolation, you reconstruct to a point not at a point. You can reconstruct at a cell, but I would say that's an awkward phrasing of entire polynomial reconstruction for a cell

File: src/Evolution/Systems/ScalarAdvection/FiniteDifference/ReconstructWork.tpp [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+  // which equals the number of interfaces.
+  const size_t reconstructed_num_pts =
+      (subcell_mesh.extents(0) + 1) *
+      subcell_mesh.extents().slice_away(0).product();
+
+  // make span for return values
```
Comment (nilsdeppe) [inline src/Evolution/Systems/ScalarAdvection/FiniteDifference/ReconstructWork.tpp]: I'm not sure this comment is illuminating without having already understood the context. Maybe something like `create views/spans into the face data, which will be filled by the reconstructor`?

File: src/Evolution/Systems/ScalarAdvection/FiniteDifference/ReconstructWork.tpp [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+    gsl::at(upper_face_vars, i) =
+        gsl::make_span(get<Tags::U>(gsl::at(*vars_on_upper_face, i))[0].data(),
+                       reconstructed_num_pts);
+    gsl::at(lower_face_vars, i) =
+        gsl::make_span(get<Tags::U>(gsl::at(*vars_on_lower_face, i))[0].data(),
+                       reconstructed_num_pts);
+  }
+
+  // number of ghost FD cells for each Direction
+  const size_t neighbor_num_pts =
```
Comment (nilsdeppe) [inline src/Evolution/Systems/ScalarAdvection/FiniteDifference/ReconstructWork.tpp]: would naming this `neighbor_num_ghost_fd_points` be clear enough to remove the comment? I prefer clearly named variables over comments when possible (not always possible, though!)

File: src/Evolution/Executables/ScalarAdvection/EvolveScalarAdvection.hpp [hunk 7]
```diff
@@ -33,6 +33,9 @@
 #include "Evolution/Systems/ScalarAdvection/BoundaryConditions/Factory.hpp"
 #include "Evolution/Systems/ScalarAdvection/BoundaryCorrections/Factory.hpp"
 #include "Evolution/Systems/ScalarAdvection/BoundaryCorrections/RegisterDerived.hpp"
+#include "Evolution/Systems/ScalarAdvection/FiniteDifference/Factory.hpp"
```
Comment (nilsdeppe) [inline src/Evolution/Executables/ScalarAdvection/EvolveScalarAdvection.hpp]: can you make switching the executable to subcell its own commit? The action list is also wrong for subcell, so I'm not sure what this part of the commit is supposed to be doing

File: tests/Unit/Evolution/Systems/ScalarAdvection/FiniteDifference/Test_MonotisedCentral.cpp [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+  auto* const mc_from_options =
+      dynamic_cast<const ScalarAdvection::fd::MonotisedCentral<Dim>*>(
+          mc_from_options_base.get());
+  REQUIRE(mc_from_options != nullptr);
+  CHECK(*mc_from_options == mc_recons);
+}
+
+}  // namespace
+
+SPECTRE_TEST_CASE("Unit.Evolution.Systems.ScalarAdvection.Fd.MonotisedCentral",
```
Comment (nilsdeppe) [inline tests/Unit/Evolution/Systems/ScalarAdvection/FiniteDifference/Test_MonotisedCentral.cpp:62]: I'm not seeing anything that actually tests the reconstruction. Did you not write a test for it?

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: src/Evolution/Executables/ScalarAdvection/EvolveScalarAdvection.hpp (hunk 7 above)
Comment (yoonso0-0) [reply src/Evolution/Executables/ScalarAdvection/EvolveScalarAdvection.hpp]: I'll include this fractional change into a separate commit for switching executable

**Comment by yoonso0-0** [comment]
@nilsdeppe Thanks for the careful review :)
My understanding on subcell code have non-linearly improved between working on this part and subcell timederivative, so I have to admit that this PR has been left somewhat unorganized and incomplete. I'll first push a fixup for `reconstructwork`, and ping you once I'm done with actual testing of MC reconstruction.

**Comment by yoonso0-0** [comment]
@nilsdeppe Pushed fixups, and added test for MC reconstruction as a separate fixup commit. Removal of noexcepts were squashed into original commits.

**Comment by nilsdeppe** [comment]
can you just squash things that add the test? I didn't review that commit to begin with

**Review by nilsdeppe — COMMENTED** [review]
LGTM, please go ahead and rebase+squash

**Review by nilsdeppe — APPROVED** [review]
LGTM.

Test failure is a timeout of `Unit.Evolution.Systems.GeneralizedHarmonic.ConstraintEnergy` in `clang-10 DEBUG, ON`
